### DISCUSSION TITLE
Why aren't you using inheritance for those structs?

### DISCUSSION THREAD

**Discussion by dhjdhj** [midi2-dev/AM_MIDI2.0Lib#3] [Ideas]
I just started looking for a C++ class library for MIDI 2.0 frankly just to save myself some time (i.e, use it before I have a deep understand of MIDI 2). 

A significant factor for me is maintenance. I started to look  at the umpProcessor.h code and noticed immediately that all three structs at the beginning share the following fields:

```
    uint8_t umpGroup;
    uint8_t messageType;
    uint8_t status;
```
so I'm left wondering why those fields are not in a "parent" struct, if you will, with the other structs inheriting from them, for example

```
struct umpCommon
{
    umpCommon() : umpGroup(255), messageType(255), status(0)  {}

    uint8_t umpGroup;
    uint8_t messageType;
    uint8_t status;

};

struct umpCVM : public umpCommon
{
    umpCVM() : channel(255),note(255), value(0),  index(0), bank(0),
         flag1(false), flag2(false) {}

    uint8_t channel;
    uint8_t note;
    uint32_t value;
    uint16_t index;
    uint8_t bank;
    bool flag1;
    bool flag2;
};

struct umpGeneric : public umpCommon
{
    umpGeneric() : value(0) {}

    uint16_t value;
};

struct umpData
{
    umpData() : streamId(0), form(0) {}

    uint8_t streamId;
    uint8_t form;
    uint8_t* data;
    uint8_t dataLength;
};
```

Alternatively, don't inherit and just have an instance the common struct in all of the other ones rather than having duplicated fields.

**Comment by starfishmod** [comment]
That sounds like a great improvement :)
I would happily accept a PR that improves the structs.

Otherwise when I do the next round of updates I'll look through and update these

**Comment by dhjdhj** [comment]
Well, it's too big a job for me - I'm not a fan of structs with default public fields that can then be modified by anybody since they're defined in a header file so I'd be changing them to classes, making the fields private, and using  getterz/setters, otherwise finding problems due to somebody changing a field directly exacerbates the maintenance problem. I was just pointing out my concerns, many may not care.

**Comment by paulhuggett** [comment]
FTR, I added PR #11 to implement this change.
